cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,329 @@
+                    output_zip_iterator,
+                    [=] __device__ (thrust::tuple<gdf_size_type, gdf_size_type> const & right_tuple,
+                                    thrust::tuple<gdf_size_type, gdf_size_type> const & left_tuple) {
+                        return ineq_op(thrust::get<1>(right_tuple), thrust::get<1>(left_tuple));
+                    });			        
+    } else {
+        auto ineq_op = row_inequality_comparator<false>(right_table, left_table, nulls_are_smallest, asc_desc.data().get()); 
+        thrust::merge(rmm::exec_policy(stream)->on(stream),
+                    left_begin_zip_iterator,
+                    left_end_zip_iterator,
+                    right_begin_zip_iterator,
+                    right_end_zip_iterator,
+                    output_zip_iterator,
+                    [=] __device__ (thrust::tuple<gdf_size_type, gdf_size_type> const & right_tuple,
+                                    thrust::tuple<gdf_size_type, gdf_size_type> const & left_tuple) {
+                        return ineq_op(thrust::get<1>(right_tuple), thrust::get<1>(left_tuple));
+                    });					        
+    }
+
+    CHECK_STREAM(stream);
+
+    return std::make_pair(out_table_indices, out_row_indices);
+}
+
+} // namespace
+
+namespace cudf {
+namespace detail {
+
+table merge(table const& left_table,
+            table const& right_table,
+            std::vector<gdf_size_type> const& key_cols,
+            std::vector<order_by_type> const& asc_desc,
+            bool nulls_are_smallest,
+            cudaStream_t stream = 0) {
+    CUDF_EXPECTS(left_table.num_columns() == right_table.num_columns(), "Mismatched number of columns");
+    if (left_table.num_columns() == 0) {
+        return cudf::empty_like(left_table);
+    }
+    
+    std::vector<gdf_dtype> left_table_dtypes = cudf::column_dtypes(left_table);
+    std::vector<gdf_dtype> right_table_dtypes = cudf::column_dtypes(right_table);
+    CUDF_EXPECTS(std::equal(left_table_dtypes.cbegin(), left_table_dtypes.cend(), right_table_dtypes.cbegin(), right_table_dtypes.cend()), "Mismatched column dtypes");
+    CUDF_EXPECTS(key_cols.size() > 0, "Empty key_cols");
+    CUDF_EXPECTS(key_cols.size() <= static_cast<size_t>(left_table.num_columns()), "Too many values in key_cols");
+    CUDF_EXPECTS(asc_desc.size() > 0, "Empty asc_desc");
+    CUDF_EXPECTS(asc_desc.size() <= static_cast<size_t>(left_table.num_columns()), "Too many values in asc_desc");
+    CUDF_EXPECTS(key_cols.size() == asc_desc.size(), "Mismatched size between key_cols and asc_desc");
+
+
+    auto gdf_col_deleter = [](gdf_column *col) {
+        gdf_column_free(col);
+        delete col;
+    };
+    using gdf_col_ptr = typename std::unique_ptr<gdf_column, decltype(gdf_col_deleter)>;
+    std::vector<gdf_col_ptr> temp_columns_to_free;
+    std::vector<gdf_column*> left_cols_sync(const_cast<gdf_column**>(left_table.begin()), const_cast<gdf_column**>(left_table.end()));
```

Comment:
Why is the `const`ness of the input being casted away? That is rarely a good idea.

Reply:
Because the `cudf::table` constructor needs `gdf_column**` or  `std::vector<gdf_column*>`.
`table.begin()` returns `gdf_column const* const*` but the underlying data is `std::vector<gdf_column*>` so it's safe to do the cast. Maybe `table.begin()` should always return `gdf_column**`?